libkbfs/block_retrieval_queue.go

```diff
@@ -72,51 +78,86 @@ func newBlockRetrievalQueue(numWorkers int) *blockRetrievalQueue {
 		ptrs:        make(map[BlockPointer]*blockRetrieval),
 		heap:        &blockRetrievalHeap{},
 		workerQueue: make(chan chan *blockRetrieval, numWorkers),
+		doneCh:      make(chan struct{}),
 	}
 }
 
 // notifyWorker notifies workers that there is a new request for processing.
 func (brq *blockRetrievalQueue) notifyWorker() {
 	go func() {
+		select {
+		case <-brq.doneCh:
+			// Prevent interference with the heap while we're retrieving from it
+			brq.mtx.Lock()
+			if brq.heap.Len() > 0 {
+				retrieval := heap.Pop(brq.heap).(*blockRetrieval)
+				// FinalizeRequest locks brq.mtx, so need to unlock it before calling
+				brq.mtx.Unlock()
+				brq.FinalizeRequest(retrieval, nil, io.EOF)
+			} else {
+				brq.mtx.Unlock()
+			}
 		// Get the next queued worker
-		ch := <-brq.workerQueue
-		// Prevent interference with the heap while we're retrieving from it
-		brq.mtx.Lock()
-		defer brq.mtx.Unlock()
-		// Pop from the heap
-		ch <- heap.Pop(brq.heap).(*blockRetrieval)
+		case ch := <-brq.workerQueue:
+			// Prevent interference with the heap while we're retrieving from it
+			brq.mtx.Lock()
+			defer brq.mtx.Unlock()
+			// Pop from the heap
+			ch <- heap.Pop(brq.heap).(*blockRetrieval)
+		}
 	}()
 }
 
 // Request submits a block request to the queue.
 func (brq *blockRetrievalQueue) Request(ctx context.Context, priority int, kmd KeyMetadata, ptr BlockPointer, block Block) <-chan error {
+	// Only continue if we haven't been shut down
+	select {
+	case <-brq.doneCh:
+		ch := make(chan error, 1)
+		ch <- io.EOF
+		return ch
+	default:
+	}
+
 	brq.mtx.Lock()
 	defer brq.mtx.Unlock()
-	br, exists := brq.ptrs[ptr]
-	if !exists {
-		// Add to the heap
-		br = &blockRetrieval{
-			blockPtr:       ptr,
-			kmd:            kmd,
-			index:          -1,
-			priority:       priority,
-			insertionOrder: brq.insertionCount,
+	// Might have to retry if the context has been canceled
+	for {
```

Comment:
The iterations limit themselves. The last statement in the loop is a `return`, and the only branch that can avoid the `return` makes it so the next iteration is guaranteed to hit it.